small dots mode (compact icons in map) (fix #4098)

moving-bits opened the pull request:
Add a switchable new mode for NewMap, in which the regular cache markers are replaced by smaller ones, to get a better readability for areas with high cache density (or lower zoom levels).

PR is based on today's (Feb 24) master branch.

moving-bits commented:
squashed into one commit

okainov commented on main/res/values/strings.xml:
```diff
@@ -1086,6 +1086,7 @@
     <string name="map_view_map">Map view</string>
     <string name="map_modes">Map settings</string>
     <string name="map_trail_show">Show trail</string>
+    <string name="map_dot_mode">Use dots</string>
```
Definitely needs rewording =) Maybe "use compact icons" or something similar

okainov commented:
I haven't tried it, but from the description can think that this can be really useful. Also would be nice of you to put screenshot to show the change visually

Also from very quick look at the code, seems you're really considering this as "dot mode". However I would think of it as of "alternative icons", which can have wider appliance (IIRC there is already some icon customization for some services)

Have you also seen\checked out the issue #5798? If you're reworking icons a bit, maybe it worth also to take a look at bigger ideas and at least make code more change-friendly or something, what do you think?

moving-bits commented:
> Also would be nice of you to put screenshot to show the change visually
Have a look at #4098 - I have a screenshot there.

> Also from very quick look at the code, seems you're really considering this as "dot mode". However I would think of it as of "alternative icons", which can have wider appliance (IIRC there is already some icon customization for some services)

> Have you also seen\checked out the issue #5798? If you're reworking icons a bit, maybe it worth also to take a look at bigger ideas and at least make code more change-friendly or something, what do you think?

Haven't seen this up to now. I took a quick glance only (I'm AFK for a few days), and from what I've seen I think those a two different things and with different motivations in mind.

#4098 is about getting a better readability / orientation on maps with a high cache density (or low zoom level), so it uses small colored dots instead of the detailed and larger icons, and without any overlays.

In contrast to this #5798 is about getting additional information displayed on the map, in this case whether a cache belongs to a certain list.

This are different things IMHO and need a different Implementation (but could be combined, if needed, so dots only with colors according to the color of a certain list).

As I said, just a quick glance, not a in-depth analysis.

okainov commented:
> those a two different things and with different motivations in mind.

Sure, it's different ideas, but what I meant is that they're related in terms of map/icons stuff. Creating an icon is mostly two steps "create an icon based on some options" + "put additional overlays on icon based on some options". 

As for alternative icons - just checked - it's for EC, you can customize either brown round icons or white rectangular. And that's what I mean - I think, the icon selection approach should be generic enough to allow flexible icons customization (at least at the moment "per cache type per service per settings value"). To be direct, in my opinion extra boolean flag is not extensible nor flexible at all, especially taking into account that it's already some kind of duplication with ExtremeCaching icon options. 
Think of someone trying to implement and add one more condition to change icon (let it be "per list" as in #5798, even though it's about overlays) - it should not require someone to go thru all the code injecting yet another parameter.

moving-bits commented:
> Sure, it's different ideas, but what I meant is that they're related in terms of map/icons stuff. Creating an icon is mostly two steps "create an icon based on some options" + "put additional overlays on icon based on some options".

That's true for today's map markers and might be true for #5798, but creating the "dot icons" for #4098 is just selecting a different set of (smaller) icons, with no overlays at all. (Another difference is that today's cache markers are relatively costly in creation due to the overlay and therefore have a special caching mechanism, whereas the dot markers are just plain and unmodified resources. So it's two pretty different ways of creating the markers which I did not want to mix.)


> To be direct, in my opinion extra boolean flag is not extensible nor flexible at all, especially taking into account that it's already some kind of duplication with ExtremeCaching icon options.

I value your direct and open response. But I've chosen this approach on purpose:
- Distinction between regular cache markers and dot marker is a user defined setting and needs to be read from the database.
- Icon creation is done somewhere deep in the calling hierarchy, and it's done frequently on a well populated map.
- So as I wanted to avoid the overhead of reading the setting for "dot mode" for each and every cache marker, I chose to read it in that place where the creation of all cache markers is done, and forward this setting using an additional parameter for the functions involved.
(If there is a less intrusive but still cost effective way to transfer this information through the class hierarchy I'm happy to use that approach.)


> Think of someone trying to implement and add one more condition to change icon (let it be "per list" as in #5798, even though it's about overlays) - it should not require someone to go thru all the code injecting yet another parameter.

If somebody's going to implement #5798 it might very well touch the cache marker creation functions (maybe even both sets of them). Then you will have to transport some more information from the controlling map routine down to the creation function (eg: is "list dot mode" activated and, if so, "which list is involved"). This could be done by switching the boolean parameter I had added into some kind of enum which represents the different combinations of settings and list ids. (You'll find this approach in other sections of the program as well.) But right now there isn't the need for this, as it's only a yes/no choice, so I chose to stick with a boolean parameter for now.

moving-bits commented:
I renamed the menu setting to "use compact icons" (thanks @okainov for the suggestion), rebased and squashed the commits into one.

Can we test drive this feature in one of our upcoming nightlies?

Lineflyer commented:
From my testing perspective I am a bit reluctant to merge this now as it involves the new map which still has some [real basic and severe bugs](https://github.com/cgeo/cgeo/issues?q=is%3Aopen+is%3Aissue+label%3A%22New+Map%22+label%3ABug). 
Packing even more features into it at this point of time might cover or change those bugs or at least it might get more complicated to get this all sorted out. On the other hand I do of course appreciate such kind of feature and we could continue fixing those new map bugs on `release` branch.
Unsure how to proceed...need to sleep some nights over it to find my opinion :)

moving-bits commented:
> From my testing perspective I am a bit reluctant to merge this now as it involves the new map which still has some [real basic and severe bugs](https://github.com/cgeo/cgeo/issues?q=is%3Aopen+is%3Aissue+label%3A%22New+Map%22+label%3ABug).
> Packing even more features into it at this point of time might cover or change those bugs or at least it might get more complicated to get this all sorted out. On the other hand I do of course appreciate such kind of feature and we could continue fixing those new map bugs on `release` branch.
> Unsure how to proceed...need to sleep some nights over it to find my opinion :)

Understood. Alas, I cannot help with those errors in NewMap (although I tried while debugging) - my knowledge of Java is not thorough enough. Probably the only ones with enough knowledge of both Java and the NewMap implementation are rsudev and pstorch.

On the other hand: Do we really want to stop all functional improvements to our maps until those bugs are fixed? IMHO the NewMap is very well usable in spite of its bugs. (I'm myself using the NewMap since a couple of months and switch back to old mapsforge map only in very rare cases, mainly when I need the red circles.) This does not mean that those bugs should not get fixed, but it should not refrain us from enhancing the functional side, IMHO.

rsudev commented:
This change is an orthogonal concern, so I would not see any issues with merging it right now. It will make addressing the open issues neither harder nor easier. I just need a good block of time to think of a different design for some parts of the new map to solve them.

Lineflyer commented:
@moving-bits Can you double confirm you consider this change ready to merge?`We just copied `master` to `release`so its the best time for new feature merging onto `master`.

moving-bits commented:
Rebased, retested (smoke test) and pushed - feel free to merge
